### PR TITLE
New function to ease interactive operations

### PR DESCRIPTION
**Pull request by francisco-dlp**
Usage
--------

```python

>>> s = hs.signals.Spectrum(np.arange(10.))
>>> ssum = hs.interactive(s.sum, axis=0)
>>> ssum.data
array(45.0)
>>> s.data /= 10
>>> s.events.data_changed.trigger()
>>> ssum.data
4.5

```

TODO
--------

- [x] unittests
- [x] User Guide documentation

NOTES
---------

Contains #423, the example is broken due to #850

### CODE REVIEW

**Comment by vidartf** [comment]
francisco-dlp/hyperspy#29

**Comment by vidartf** [comment]
No this one has the broken CI as well, no idea why, but feel free to rebase as much as needed, I have no code based on this/events PR that I cannot rebase again.

File: hyperspy/interactive.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        # To only trigger once even with several changes, we suppress here
+        # and trigger manually below if there were any changes.
+        changes = False
+        with self.out.axes_manager.events.suppress():
+            for ax_src, ax_dst in zip(out.axes_manager._axes,
+                                      self.out.axes_manager._axes):
+                c = ax_dst.update_from(ax_src, ('offset', 'scale', 'size'))
+                changes = changes or c
+        if changes:
+            self.out.axes_manager.events.transformed.trigger(self.out)
```
Comment (to266) [inline hyperspy/interactive.py]: The trigger argument has to be `AxesManager`, not `Signal`, right?

Comment (vidartf) [inline hyperspy/interactive.py]: Yes, should be `self.out.axes_manager`. Well spotted!

**Comment by vidartf** [comment]
@francisco-dlp : Would you mind merging the latest version of #423 into this? There should be some minor issues to fix in the merge, but it should also fix CI. Then I can merge this into my PR that depend on it.

**Comment by francisco-dlp** [comment]
It'll have to wait until tomorrow. I have to leave the office and there are two non trivial failures two fix.

**Comment by francisco-dlp** [comment]
I've just merged it.

**Comment by francisco-dlp** [comment]
@vidartf, what's the point of the `AxesManager` `transformed` event? It is introduced in this branch but not used. Beside, wouldn't tit be more useful to define wo `DataAxis` events, `axis_data_changed` and `axis_size_changed` instead?

**Comment by vidartf** [comment]
`transformed` is currently used in `Interactive._recompute_out()`. The purpose of it is to have a single event when the space spanned by the axes manager transforms. You could have the other events too, but this event should hopefully only trigger once even if several axes are updated. In the same ways, new widgets have events `moved`, `resized` and `changed`, where the last one triggers when it's either moved or resized, but only once when it is both moved and resized.

**Comment by francisco-dlp** [comment]
This one should be ready for review.

File: hyperspy/tests/test_interactive.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        s = self.s
+        e1, e2 = Event(), Event()
+        ss = hs.interactive(s.sum, e1, axis=0)
+        sss = hs.interactive(ss.sum, e2, axis=0)
+        np.testing.assert_allclose(sss.data, np.sum(s.data, axis=(0, 1)))
+        s.data += 3.2
+        nt.assert_false(np.allclose(ss.data, np.sum(s.data, axis=(1))))
+        e1.trigger()
+        np.testing.assert_allclose(ss.data, np.sum(s.data, axis=(1)))
+        nt.assert_false(np.allclose(ss.data, np.sum(s.data, axis=(0, 1))))
```
Comment (to266) [inline hyperspy/tests/test_interactive.py]: should be `sss`

Comment (francisco-dlp) [inline hyperspy/tests/test_interactive.py]: Fixed, thanks!

**Comment by to266** [comment]
...Is there a way to force github update the diff? I've looked at it locally, but would like to make sure

**Comment by vidartf** [comment]
It's not unlikely that there has been some rebasing or merge conflicts somewhere. The diff could therefore be artificially large until master is merged again.

File: doc/user_guide/tools.rst [hunk 3]
```diff
@@ -951,3 +952,48 @@ operation.
...
+
+Interactive operations
+----------------------
+
+.. versionadded:: 0.9
+
+
+The function :py:func:`~.interactive.interactive` ease the task of defining
+operations that are automatically updated when an event is triggered. By
+default is recomputes the operation when data or the axes of the original
```
Comment (to266) [inline doc/user_guide/tools.rst]: ...default __it__ recomputes..

Comment (francisco-dlp) [inline doc/user_guide/tools.rst]: Fixed, thanks.

**Pull request merged**